### ISSUE TITLE
Add scratch and fibre removal as merging from both sides of a line (polygon/spline)

### ISSUE THREAD

**Issue by sobukus** [Beep6581/RawTherapee#6419]
I am exploring negative scanning via DSLR and RawTherapee and am getting quite nice results already.
I really like being able to program any processing in the non-destructive sidecar files, avoiding costly
intermediate bitmaps. Ideally, that would extend to some operations on masked areas, or actual drawing
operations. Well, the vignette filter is already such an operation … anyhow. Back to my suggestion about
enhancing the dirt removal on negative scans (or actually, dirty DSLR sensors or lenses, too).

I yet don't have a very clean environment and checked out the spot removal tool to quite some excess to
get a clean picture  — around 80 spots on one image in one serious instance.

Of course, properly cleaning the film beforehand is preferable, but you don't always have the time or
correct tools. For example, I got some persistent bit of fibre on one negative that I mistook for a scratch
the first time, alerting me to the possibility of causing scratches by too forceful cleaning.

That fibre results in a quite straight white line of about 2 to 3 pixels width in the resulting positive image
from a 12 MP RAW file. Other small curled fibres also appear with less thickness, even. I did spend some
dozen spots on those (including reducing the minimal radius of those, see my pull request), tracking the
shape to finally hide it. This works, but it is more destructive to the local detail than it has to be.

I figured that a better approach for this kind of very thin elongated disturbances would be to fit a curve
along some control points and then have pixels mixed in from the left and right sides instead of
searching for a source area elsewhere in the image. Care would need to be taken to keep the pixel
grain/noise of the original material while keeping things smooth, just like with the spots doing
that cleverly by blending in areas.

I wonder how much of the spot approach could be re-used/adapted. I imagine this approach of just
filling in from the direct neighbours would apply to any spot in the size of a few pixels only. The
elongated scratch or even curled fibre would be an application of the same principle on each pixel
around the line.

Was something along that considered/tried/rejected already? If someone could get me started with
some UI code for the curve drawing part for a new  tiny spot / thin scratch or fible tool in the code base,
I could go some way trying algorithms. But maybe this is already rather obvious for the person(s)
working on the spot tool. And I am quite certain that there is prior art on this.

If we got spots and scratches covered in RT 5.9, that would be splendit! Just in time for me discovering
my old analogue photos and that there are pictures in the negative film that I didn't even see developed
yet;-)

**Comment by Lawrence37** [comment]
The spot removal tool is simple in principle. Each spot is a circle (easy to compute) and pixels are simply cloned from the source to destination. The source location is determined by user input.

Using paths or polygons makes things a bit more complicated because you need to determine which pixels are near the path or inside the polygon. It would be wise to develop such a feature for general use so that it can be incorporated into local adjustments, for example.

The method of sampling pixels from either sides of a line sounds like inpainting. If there are going to be multiple shapes and cloning/healing methods, we should be careful about keeping the user interface simple and intuitive. A user could be confused why spot sources are manually specified while paths are automatically inpainted.

**Comment by sobukus** [comment]
Yes. Indeed I also have a suggestion to make the spot tool itself less simple;-) There are cases where it would really help to be able to define a rotation. I worked on a negative scan with a tree in winter. Lots of branches in differing angles. When I need to cover a certain spot touching (a) branch(es), I need to search for similar features elsewhere, also in the correct angle. I see how that could compligate the UI further, but a rotation handle might be doable (it can be well outside the circles).

I see how you want to have RT still usable by non-programmers, while there is all that possibility of getting scripted edits of endless complexity in. I actually think a separate tab with editing operations would be nice … basically integrating GIMP without layers with choice on when in the processing stack those operations happen. I can see why one in the end would desist from that. Staying careful at the boundary to actual image drawing with the spot removal.

So, back to the points. I do think that both approaches are valuable. My ideal RT has two sub-tools

1. spots with copy-from location and rotation
2. paths with defined width for inpainting

A path could be a single point, where inpainting would happen from all directions, I guess. So that would be the second way to treat a single spot. Either copy from somewhere else or use the direct surroundings. This would be valuable and the only one where one realistically could add automatic spot detection and removal (with added IR channel, perhaps).

We could keep the shape at a simple path of connected dots, even just linear. If you want more smoothness, add more dots. I guess that simplifies the math considerably.

UI-wise, my suggestion thus would result in a second section in the spot removal tab, or a new little section called scratch removal. Is there already UI to draw paths at all? I guess we could re-use and extend the spot drawing.

1. Have a spot selected.
2. Press and hold some other key (shift?).
3. Click someplace to add a dot that is connected to the erstwhile selected one.
4. You got a path now.

The circle UI is still useful. The main circle size would be set to match the width of the scratch/dot to cover, the feather to set the width of the area for doing inpainting from. The added angle control could be useful to define a single direction to draw from. But this is not really that useful, as once you got a path, you can use the perpendicular direction to draw across it. But one could keep things simple for single-point paths, where the angle would be needed to be defined if we do not have a circular inpainting mode for endpoints or single spots. I do think, though, that we should have that circular mode for single points and path endpoints, maybe even for vertices in the path, to have things smoother.

So the UI would borrow a lot from the current spot UI, including moving a point, but keeping its connections. I feel confident that I could come up with the algorithmic part of the painting, testing it with some of my dirty examples. I'd need to do quite a bit of reading to get along in the RT UI code, though.

I guess having some kind of per-spot switch about the painting mode could be overkill. There's really minimal screen estate being used for the spot removal with a simple count of spots and a button to activate the editing. I suggest a second line for paths with its own edit button and tooltip help text.

I do not imagine drawing polygons with inside/outside areas (yet). I only talk about non-closed paths here with a simple algorithm that only needs to know which points it is between and work out linear interpolation of parameters (widths) and the perpendicular direction. On the points themselves, a rounded inpainting mode might be needed.

I am not sure yet how complicated the painting algorithm needs to be for thin scratches. Simple linear interpolation across the gap might work, but maybe some higher order is needed to hide things properly. Working pixel per pixel could be expensive, but necessary when we also interpolate the width between the two segment points. It could be a lot more elaborate than the current spot tool. My aim would be that a handful of lines of some hundreds of pixels length each does not increase preview times much, but of course you could always draw so many lines that the CPU will be sweating for those.

In any case, I do think that keeping only two options  in the UI (cloning for sponts, inplainting for paths, with single-point paths as special case that doesn't need extra UI) could be all that is really needed. I briefly wondered about clone mode for paths, but this should be rarely useful without clearly visible artifacts.

**Comment by sobukus** [comment]
Another idea … inpainting would be one thing, but another would be a bit of re-interpolation of the surroundings to stretch them, closing a gap. You got a wound and you close it by pulling together the skin around it. But I figure that's a lot more complicated.
